### PR TITLE
Add option to collect must-gather

### PR DESCRIPTION
**Pull request by rnetser**

### CODE REVIEW

File: openshift_cli_installer/cli.py [hunk 1]
```diff
@@ -194,6 +194,15 @@
 """,
     type=click.Path(exists=True),
 )
+@click.option(
+    "--must-gather-output-dir",
+    help="""
+\b
+Path to must-gather output directory.
+must-gather will be collected when cluster installation fails.
```
Comment (myakove) [inline openshift_cli_installer/cli.py:202]: will try to collect (if we have a cluster)

File: openshift_cli_installer/cli.py [hunk 2]
```diff
@@ -194,6 +194,15 @@
...
+""",
+    type=click.Path(exists=True),
```
Comment (myakove) [inline openshift_cli_installer/cli.py:204]: Why `exists=True`? 
I think the user can pass `/tmp/my-must-gatther-dir` and we will create the dir.

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py [hunk 3]
```diff
@@ -62,15 +64,22 @@ def osd_create_cluster(cluster_data):
...
-            f"Failed to run cluster create for cluster {cluster_data['name']}\n{ex}",
+            f"Failed to run cluster create for cluster {cluster_name}\n{ex}",
             fg=ERROR_LOG_COLOR,
         )
 
+        if must_gather_output_dir:
+            collect_must_gather(
+                must_gather_output_dir=must_gather_output_dir,
+                cluster_data=cluster_data,
+                cluster_object=cluster_object,
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/osd_clusters.py:80]: don't we already have `cluster_object` in `cluster_data`?

File: openshift_cli_installer/utils/clusters.py [hunk 4]
```diff
@@ -118,15 +120,16 @@ def add_cluster_info_to_cluster_data(cluster_data, cluster_object=None):
...
+def set_cluster_auth(cluster_data, cluster_object, write_password_file=True):
     auth_path = os.path.join(cluster_data["install-dir"], "auth")
     Path(auth_path).mkdir(parents=True, exist_ok=True)
 
     with open(os.path.join(auth_path, "kubeconfig"), "w") as fd:
         fd.write(yaml.dump(cluster_object.kubeconfig))
 
-    with open(os.path.join(auth_path, "kubeadmin-password"), "w") as fd:
-        fd.write(cluster_object.kubeadmin_password)
+    if write_password_file:
```
Comment (myakove) [inline openshift_cli_installer/utils/clusters.py:130]: Why this 'if' needed?

File: openshift_cli_installer/utils/clusters.py [hunk 5]
```diff
@@ -162,3 +165,40 @@ def add_s3_bucket_data(clusters, s3_bucket_name, s3_bucket_path=None):
         )
 
     return clusters
+
+
+def collect_must_gather(must_gather_output_dir, cluster_data, cluster_object=None):
+    name = cluster_data["name"]
```
Comment (myakove) [inline openshift_cli_installer/utils/clusters.py:171]: everything needed to go under the `try`, we don't want to fail here never (we cal destroy after)

**Review by myakove — CHANGES_REQUESTED** [review]

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py (hunk 3 above)
Comment (rnetser) [reply openshift_cli_installer/libs/managed_clusters/osd_clusters.py:80]: yes; moved the addition of it to earlier in the flow (till now it was after cluster create)

File: openshift_cli_installer/utils/clusters.py (hunk 4 above)
Comment (rnetser) [reply openshift_cli_installer/utils/clusters.py:130]: removed

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py [hunk 6]
```diff
@@ -22,6 +22,8 @@ def osd_create_cluster(cluster_data, must_gather_output_dir=None):
         client=cluster_data["ocm-client"],
         name=cluster_name,
     )
+    cluster_data["cluster-object"] = cluster_object
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/osd_clusters.py:25]: We need to set `cluster_data["cluster-object"]` in the same place we do `cluster_data["ocm-client"]` `openshift_cli_installer.utils.cli_utils.prepare_clusters`

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py [hunk 7]
```diff
@@ -72,12 +73,12 @@ def osd_create_cluster(cluster_data, must_gather_output_dir=None):
             f"Failed to run cluster create for cluster {cluster_name}\n{ex}",
             fg=ERROR_LOG_COLOR,
         )
+        set_cluster_auth(cluster_data=cluster_data, cluster_object=cluster_object)
```
Comment (myakove) [inline openshift_cli_installer/libs/managed_clusters/osd_clusters.py:76]: Now `cluster_data` contain `cluster_object`, no need to pass it to `set_cluster_auth`

File: openshift_cli_installer/utils/clusters.py [hunk 8]
```diff
@@ -167,26 +166,36 @@ def add_s3_bucket_data(clusters, s3_bucket_name, s3_bucket_path=None):
...
-def collect_must_gather(must_gather_output_dir, cluster_data, cluster_object=None):
-    name = cluster_data["name"]
-    platform = cluster_data["platform"]
-    target_dir = os.path.join(must_gather_output_dir, "must-gather", platform, name)
+def collect_must_gather(must_gather_output_dir, cluster_data):
+    try:
+        name = cluster_data["name"]
+        platform = cluster_data["platform"]
+        target_dir = os.path.join(must_gather_output_dir, "must-gather", platform, name)
+    except Exception:
```
Comment (myakove) [inline openshift_cli_installer/utils/clusters.py:174]: we should also log the exception.

File: openshift_cli_installer/utils/clusters.py [hunk 9]
```diff
@@ -167,26 +166,36 @@ def add_s3_bucket_data(clusters, s3_bucket_name, s3_bucket_path=None):
...
+        kubeconfig_path = get_kubeconfig_path(cluster_data=cluster_data)
+        if not os.path.exists(kubeconfig_path):
+            click.secho(
+                f"kubeconfig for cluster {name} does not exist; cannot run"
+                " must-gather.",
+                fg=ERROR_LOG_COLOR,
+            )
+            return
+
         click.echo(f"Prepare target extracted directory {target_dir}.")
```
Comment (myakove) [inline openshift_cli_installer/utils/clusters.py:192]: add must-gather reference to the error log

**Review by myakove — CHANGES_REQUESTED** [review]

File: openshift_cli_installer/libs/managed_clusters/osd_clusters.py (hunk 7 above)
Comment (rnetser) [reply openshift_cli_installer/libs/managed_clusters/osd_clusters.py:76]: yes we do; we need the file

Comment (myakove) [reply openshift_cli_installer/libs/managed_clusters/osd_clusters.py:76]: Once you set it in `openshift_cli_installer.utils.cli_utils.prepare_clusters` you can take it from `cluster_data`

Comment (rnetser) [reply openshift_cli_installer/libs/managed_clusters/osd_clusters.py:76]: ahhh

File: openshift_cli_installer/cli.py [hunk 10]
```diff
@@ -194,7 +194,16 @@
 """,
     type=click.Path(exists=True),
 )
-def main(**kwargs):
+@click.option(
+    "--must-gather-output-dir",
+    help="""
+\b
+Path to must-gather output directory.
+must-gather will try be collect data when cluster installation fails and cluster can be accessed.
```
Comment (omrirh) [inline openshift_cli_installer/cli.py:202]:
```suggestion
must-gather will try to collect data when cluster installation fails and cluster can be accessed.
```

**Review by myakove — CHANGES_REQUESTED** [review]

**Review by myakove — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Pull request merged**